<a href="https://colab.research.google.com/github/JSEFERINO/C_Datos/blob/main/15_C_Datos_Data_Perdidos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Curso Introductorio a la Ciencia de Datos por Julio Hurtado Márquez**
## **Parte V**
# **70. Importar y exportar dataframes limpios**

## **Acerca de los Datos**
- Estos datos abordan el rendimiento de los estudiantes en la educación secundaria de dos escuelas portuguesas. Los atributos de los datos incluyen calificaciones de los estudiantes, características demográficas, sociales y relacionadas con la escuela) y se recopilaron mediante el uso de informes y cuestionarios escolares. Se proporcionan dos conjuntos de datos sobre el rendimiento en dos materias distintas: Matemáticas (mat) y lengua portuguesa (por). En [Cortez and Silva, 2008], los dos conjuntos de datos se modelaron bajo tareas de regresión y clasificación binaria/de cinco niveles. Nota importante: el atributo objetivo G3 tiene una fuerte correlación con los atributos G2 y G1. Esto ocurre porque G3 es la calificación del último año (emitida en el 3er período), mientras que G1 y G2 corresponden a las calificaciones del 1er y 2do período. Es más difícil predecir G3 sin G2 y G1, pero dicha predicción es mucho más útil (consulte la fuente en papel para obtener más detalles).

## **Attribute Information**
### Attributes for both student-mat.csv (Math course) and student-por.csv (Portuguese language course) datasets:
- 1 school - escuela de estudiantes (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira)
- 2 sex - student's sex (binary: 'F' - female or 'M' - male)
- 3 age - student's age (numeric: from 15 to 22)
- 4 address - student's home address type (binary: 'U' - urban or 'R' - rural)
- 5 famsize - family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3)
- 6 Pstatus - parent's cohabitation status (binary: 'T' - living together or 'A' - apart)
- 7 Medu - mother's education (numeric: 0 - none,  1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
- 8 Fedu - father's education (numeric: 0 - none,  1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
- 9 Mjob - mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
- 10 Fjob - father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
- 11 reason - reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other')
- 12 guardian - student's guardian (nominal: 'mother', 'father' or 'other')
- 13 traveltime - home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour)
- 14 studytime - weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours)
- 15 failures - number of past class failures (numeric: n if 1<=n<3, else 4)
- 16 schoolsup - extra educational support (binary: yes or no)
- 17 famsup - family educational support (binary: yes or no)
- 18 paid - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no)
- 19 activities - extra-curricular activities (binary: yes or no)
- 20 nursery - attended nursery school (binary: yes or no)
- 21 higher - wants to take higher education (binary: yes or no)
- 22 internet - Internet access at home (binary: yes or no)
- 23 romantic - with a romantic relationship (binary: yes or no)
- 24 famrel - quality of family relationships (numeric: from 1 - very bad to 5  excellent)
- 25 freetime - free time after school (numeric: from 1 - very low to 5 - very high)
- 26 goout - going out with friends (numeric: from 1 - very low to 5 - very high)
- 27 Dalc - workday alcohol consumption (numeric: from 1 - very low to 5 - very high)
- 28 Walc - weekend alcohol consumption (numeric: from 1 - very low to 5 - very high)
- 29 health - current health status (numeric: from 1 - very bad to 5 - very good)
- 30 absences - number of school absences (numeric: from 0 to 93)

##  these grades are related with the course subject, Math or Portuguese:

- 31 G1 - first period grade (numeric: from 0 to 20)
- 31 G2 - second period grade (numeric: from 0 to 20)
- 32 G3 - final grade (numeric: from 0 to 20, output target)



## **a. Nuestras librerias.**

In [ ]:
## ** a. Nuestras librerias.**
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## **b. Importamos la Base de datos con la cual vamos a trabajar**
- https://archive-beta.ics.uci.edu/dataset/320/student+performance
- https://archive.ics.uci.edu/ml/machine-learning-databases/00320/

In [ ]:
## **b. Importamos la Base de datos con la cual vamos a trabajar**
import cv2 as cv
from google.colab import files
uploaded=files.upload()

Saving student-mat.csv to student-mat.csv


## **c. El DataFrame de la Base de datos con la cual vamos a trabajar**

In [ ]:
## **c. El DataFrame de la Base de datos con la cual vamos a trabajar**
data=pd.read_csv("/content/student-mat.csv",delimiter=';')
data

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

## **d. Que tipos de datos tiene la base de datos a trabajar**

- Predecir el rendimiento de los estudiantes en la educación secundaria (bachillerato).

|Características del conjunto de datos|multivariado|
|:--:|:--:|
|Área temática|Social|
|Tareas asociadas|Clasificación, Regresión|
|Tipo de atributo|Entero|
|Instancias|395|
|Atributos|33|

## **e. Tipos de variables de nuestro DataFrame**

In [ ]:
## **e. Tipos de variables de nuestro DataFrame**
data.dtypes

school        object
sex           object
age            int64
address       object
famsize       object
Pstatus       object
Medu           int64
Fedu           int64
Mjob          object
Fjob          object
reason        object
guardian      object
traveltime     int64
studytime      int64
failures       int64
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
G1             int64
G2             int64
G3             int64
dtype: object

# **71. Trabajando con nuestra base de Datos**

#**a. Vamos a seleccionar solo los datos numericos de nuestra base de datos**

In [ ]:
#**a. Vamos a seleccionar solo los datos numericos de nuestra base de datos**
data0 = data.select_dtypes(include='int64')
data0
data1 = data0.copy()

## **b. Ahora guardemos la data original en una copia**

In [ ]:
## **b. Ahora guardemos la data original en una copia**
data_ori = data
data_ori.head(10)

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4   at_home   teacher   
1     GP   F   17       U     GT3       T     1     1   at_home     other   
2     GP   F   15       U     LE3       T     1     1   at_home     other   
3     GP   F   15       U     GT3       T     4     2    health  services   
4     GP   F   16       U     GT3       T     3     3     other     other   
5     GP   M   16       U     LE3       T     4     3  services     other   
6     GP   M   16       U     LE3       T     2     2     other     other   
7     GP   F   17       U     GT3       A     4     4     other   teacher   
8     GP   M   15       U     LE3       A     3     2  services     other   
9     GP   M   15       U     GT3       T     3     4     other     other   

   ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0  ...      4        3      4     1     1      3        6   5   6   6  
1  ...      5        3      3     1     1      3        4   5   5   6  
2  ...      4        3      2     2     3      3       10   7   8  10  
3  ...      3        2      2     1     1      5        2  15  14  15  
4  ...      4        3      2     1     2      5        4   6  10  10  
5  ...      5        4      2     1     2      5       10  15  15  15  
6  ...      4        4      4     1     1      3        0  12  12  11  
7  ...      4        1      4     1     1      1        6   6   5   6  
8  ...      4        2      2     1     1      1        0  16  18  19  
9  ...      5        5      1     1     1      5        0  14  15  15  

[10 rows x 33 columns]

# **72. Cómo encontrar valores NaN en un DataFrame y modificarlos**

- Cuando trabajamos con un DataFrame, especialmente si es extenso, podemos tener problemas con lo valores NaN. 
- Los NaN (“Not a Number“) son valores vacíos no computables que debemos tratar en nuestros conjuntos de datos antes de trabajar con ellos.
- Veremos algunos códigos de Python que pueden ayudarnos a trabajar con los NaN.
- Este código nos devolverá True si hay algún valor NaN en nuestro DataFrame.

## **a. Será que nuestra base de datos tiene datos perdidos**

In [ ]:
## **a. Será que nuestra base de datos tiene datos perdidos**
data.isnull().values.any()

False

## **b. ¿Cuántos NaN hay?**

In [ ]:
## **b. ¿Cuántos NaN hay?**
data[data.isna().any(axis=1)] # Ninguno

Empty DataFrame
Columns: [school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3]
Index: []

[0 rows x 33 columns]

## **c. Colocaremos la mayor cantidad de NaN excepto en las tres últimas columnas de la data1**
- Para no hacerlo manual hagamos, usemos un bucle _for_**

In [ ]:
## **c. Colocaremos la mayor cantidad de NaN excepto en las tres últimas columnas de la data1**
length = len(data1) # En toda la longitud de la base de datos
for i in data1.columns[:-3]: # Para cualquier columna desde la primera hasta la antepenultima
    num = int(np.random.randint(0,90)/100*length) #Colocaremos un número aleatorio de la dist uniforme entre 0 y 90 en toda 
    # la longitud
    idx_replace = np.random.randint(0, length-1, num) # despues a ese numero lo reemplazamos por cero
    data1.loc[idx_replace, i] = np.nan # al final lo reemplazamos pon NaN

## **d. Ahora veamos como quedó nuestra base de datos**

In [ ]:
## **d. Ahora veamos como quedó nuestra base de datos**
data1

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2     NaN   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4     NaN   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  

[395 rows x 16 columns]

## **e. Nuevamente le preguntamos a la base de datos si presenta NaN**

In [ ]:
## **e. Nuevamente le preguntamos a la base de datos si presenta NaN**
data1.isnull().values.any()

True

## **f. ¿Cuántos NaN hay ahora?**

In [ ]:
## **f. ¿Cuántos NaN hay ahora?**
data1[data1.isna().any(axis=1)] # Ninguno

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2     NaN   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4     NaN   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
389  18.0   1.0   NaN         NaN        2.0       NaN     1.0       1.0   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
389    NaN   NaN   1.0     5.0       0.0   6   5   0  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  

[391 rows x 16 columns]

##**g. Ahora dime donde se encuentran los datos faltantes**
- Que forma tan difícil de preguntar

In [ ]:
##**g. Ahora dime donde se encuentran los datos faltantes**
for i in data1.columns:
    print("Hay {} ({}%) datos faltantes en la columna {}"
    .format(len(data1[data1[i].isna()]), np.round(len(data1[data1[i].isna()])/len(data1)*100,1), i))


Hay 136 (34.4%) datos faltantes en la columna age
Hay 115 (29.1%) datos faltantes en la columna Medu
Hay 163 (41.3%) datos faltantes en la columna Fedu
Hay 177 (44.8%) datos faltantes en la columna traveltime
Hay 192 (48.6%) datos faltantes en la columna studytime
Hay 243 (61.5%) datos faltantes en la columna failures
Hay 22 (5.6%) datos faltantes en la columna famrel
Hay 149 (37.7%) datos faltantes en la columna freetime
Hay 187 (47.3%) datos faltantes en la columna goout
Hay 130 (32.9%) datos faltantes en la columna Dalc
Hay 139 (35.2%) datos faltantes en la columna Walc
Hay 19 (4.8%) datos faltantes en la columna health
Hay 52 (13.2%) datos faltantes en la columna absences
Hay 0 (0.0%) datos faltantes en la columna G1
Hay 0 (0.0%) datos faltantes en la columna G2
Hay 0 (0.0%) datos faltantes en la columna G3


##**h. Pero hay una forma mas sencilla de preguntar**

In [ ]:
##**h. Pero hay una forma mas sencilla de preguntar**
data1.isnull().sum()

age           136
Medu          115
Fedu          163
traveltime    177
studytime     192
failures      243
famrel         22
freetime      149
goout         187
Dalc          130
Walc          139
health         19
absences       52
G1              0
G2              0
G3              0
dtype: int64

#**73. Analicemos cuales son los Tipos de datos perdidos**

#### **Datos perdidos completamente al azar (Missing Completely at Random, MCAR):**

 - - Missing Completely at Random (MCAR) es bastante sencillo. Lo que significa es lo que dice: _la propensión a que un punto de datos falte es completamente aleatoria._
 - - No hay ninguna relación entre el hecho de que un punto de datos falte y cualquier valor del conjunto de datos, faltante u observado.

#### **Falta no aleatoria (Missing not at random, MNAR):**

- - También conocida como falta de respuesta no asignable, son datos que no son ni MAR ni MCAR (es decir, el valor de la variable que falta está relacionado con la razón por la que falta).


#### **Falta al azar (Missing at Random, MAR):**

 - - Significa que la propensión a que un punto de datos falte no está relacionada con los datos que faltan, pero sí con algunos de los datos observados. 
 - - El hecho de que alguien haya respondido al número 13 en su encuesta no tiene nada que ver con los valores que faltan, pero sí con los valores de alguna otra variable.

In [ ]:
## **a. Volvamos a nuestra base de datos**
data1

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2     NaN   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4     NaN   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  

[395 rows x 16 columns]

#**74. Todas las técnicas para manejar valores perdidos**

- **Reemplazo de media/mediana/moda**
- **Imputación de muestra aleatoria**
- **Captura de valores NAN con una nueva característica**
- **Imputación de fin de distribución**
- **Imputación arbitraria**
- **Imputación de Categorías Frecuentes**

## **a. Eliminado las filas que contengan datos faltantes**

In [ ]:
## **a. Eliminado las filas que contengan datos faltantes**
data2=data1.copy()
data2

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2     NaN   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4     NaN   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  

[395 rows x 16 columns]

##**b. Esta es la cantidad de datos faltantes**

In [ ]:
##**b. Esta es la cantidad de datos faltantes**
data2.isnull().sum()

age           136
Medu          115
Fedu          163
traveltime    177
studytime     192
failures      243
famrel         22
freetime      149
goout         187
Dalc          130
Walc          139
health         19
absences       52
G1              0
G2              0
G3              0
dtype: int64

## **c. Eliminando las filas con con datos faltantes**

In [ ]:
## **c. Eliminando las filas con con datos faltantes**
print(data2.dropna())

      age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
10   15.0   4.0   4.0         1.0        2.0       0.0     3.0       3.0   
95   15.0   1.0   1.0         2.0        4.0       1.0     3.0       1.0   
275  17.0   2.0   2.0         2.0        2.0       0.0     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
10     3.0   1.0   2.0     2.0       0.0  10   8   9  
95     2.0   1.0   1.0     1.0       2.0   7  10  10  
275    4.0   2.0   3.0     5.0       6.0  12  12  12  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  


In [ ]:
print(data2.dropna().isnull().sum())

age           0
Medu          0
Fedu          0
traveltime    0
studytime     0
failures      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64


## **d. Hagamos un resumen de los datos**

In [ ]:
## **d. Hagamos un resumen de los datos**
data1.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  259.000000  280.000000  232.000000  218.000000  203.000000  152.000000   
mean    16.729730    2.700000    2.551724    1.481651    1.980296    0.348684   
std      1.362682    1.045214    1.079882    0.732481    0.801981    0.774253   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    3.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  373.000000  246.000000  208.000000  265.000000  256.000000  376.000000   
mean     3.941019    3.272358    3.081731    1.543396    2.269531    3.550532   
std      0.890062    1.015697    1.048603    0.932882    1.277869    1.384716   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  343.000000  395.000000  395.000000  395.000000  
mean     5.810496   10.908861   10.713924   10.415190  
std      8.309208    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

## **e. Nuevamente los datos**

In [ ]:
## **e. Nuevamente los datos**
data3 = data1.copy()
data3

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2     NaN   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4     NaN   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  

[395 rows x 16 columns]

## **f. Hemos imputado los NaN por la media en la data 3**

In [ ]:
## **f. Hemos imputado los NaN por la media en la data 3**
data3['age'] = data3['age'].replace(np.nan,17)
data3

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2    17.0   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4    17.0   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  

[395 rows x 16 columns]

## **g. Verificamos que la variable 'age' no tiene datos faltantes**

In [ ]:
## **g. Verificamos que la variable 'age' no tiene datos faltantes**
data3.isnull().sum()

age             0
Medu          115
Fedu          163
traveltime    177
studytime     192
failures      243
famrel         22
freetime      149
goout         187
Dalc          130
Walc          139
health         19
absences       52
G1              0
G2              0
G3              0
dtype: int64

#**75. Todas las técnicas para manejar valores perdidos - Parte 2**

## **a. Dado que hemos modificado nuestra data, hagamos una nueva copia de la misma**

In [ ]:
## **a. Dado que hemos modificado nuestra data, hagamos una nueva copia de la misma**
data3 

age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   NaN   4.0         NaN        NaN       NaN     4.0       NaN   
1    17.0   NaN   NaN         NaN        NaN       NaN     5.0       NaN   
2    17.0   1.0   NaN         NaN        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         NaN        NaN       0.0     3.0       NaN   
4    17.0   3.0   3.0         NaN        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   NaN         1.0        2.0       2.0     NaN       NaN   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   NaN   1.0         NaN        NaN       NaN     5.0       5.0   
393  18.0   NaN   NaN         NaN        NaN       NaN     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences  G1  G2  G3  
0      4.0   1.0   1.0     3.0       6.0   5   6   6  
1      3.0   1.0   1.0     3.0       4.0   5   5   6  
2      NaN   NaN   3.0     3.0      10.0   7   8  10  
3      NaN   1.0   1.0     5.0       2.0  15  14  15  
4      2.0   1.0   2.0     5.0       4.0   6  10  10  
..     ...   ...   ...     ...       ...  ..  ..  ..  
390    4.0   4.0   NaN     4.0      11.0   9   9   9  
391    NaN   3.0   4.0     2.0       3.0  14  16  16  
392    3.0   3.0   NaN     3.0       3.0  10   8   7  
393    1.0   3.0   NaN     5.0       0.0  11  12  10  
394    3.0   3.0   3.0     5.0       5.0   8   9   9  

[395 rows x 16 columns]

# **76. No todas las variables se pueden imputar con la media - como hacemos?**


- 1 school  - escuela de estudiantes (binario: 'GP' - Gabriel Pereira o 'MS' - Mousinho da Silveira)
- 2 sex - sexo del estudiante (binario: 'F' - femenino o 'M' - masculino)
- 3 age  - edad del estudiante (numérico: de 15 a 22)
- 4 address  - tipo de domicilio del estudiante (binario: 'U' - urbano o 'R' - rural)
- 5 famsize - tamaño de familia (binario: 'LE3' - menor o igual a 3 o 'GT3' - mayor a 3)
- 6 Pstatus - estado de convivencia de los padres (binario: 'T' - viviendo juntos o 'A' - separados)
- 7 Medu - educación de la madre (numérico: 0 - ninguno, 1 - educación primaria (4º grado), 2 - 5º a 9º grado, 3 - educación secundaria o 4 - educación superior)
- 8 Fedu - educación del padre (numérico: 0 - ninguno, 1 - educación primaria (4° grado), 2 – 5° a 9° grado, 3 – educación secundaria o 4 – educación superior)
- 9 Mjob - trabajo de la madre (nominal: 'maestra', 'salud' relacionado con la atención, 'servicios' civiles (por ejemplo, administrativo o policía), 'en_casa' u 'otro')
- 10 Fjob - trabajo del padre (nominal: 'maestro', 'salud' relacionado con la atención, 'servicios' civiles (por ejemplo, administrativo o policía), 'en_casa' u 'otro')
- 11 reason  - razón para elegir esta escuela (nominal: cerca de 'casa', 'reputación' de la escuela, preferencia de 'curso' u 'otro')
- 12 tutor - tutor del estudiante (nominal: 'madre', 'padre' u 'otro')
- 13 traveltime - tiempo de viaje de casa a la escuela (numérico: 1 - <15 min., 2 - 15 a 30 min., 3 - 30 min. a 1 hora, o 4 - >1 hora)
- 14 studytime  - tiempo de estudio semanal (numérico: 1 - <2 horas, 2 - 2 a 5 horas, 3 - 5 a 10 horas, o 4 - >10 horas)
- 15 failures  - número de fallas de clase pasadas (numérico: n si 1<=n<3, si no 4)
- 16 schoolsup  - apoyo educativo extra (binario: si o no)
- 17 famsup - apoyo educativo familiar (binario: si o no)
- 18 paid  pagas extra pagadas dentro de la materia del curso (matemáticas o portugués) (binario: sí o no)
- 19 activities  - actividades extraescolares (binario: sí o no)
- 20 nursery  - parvulario asistido (binario: sí o no)
- 21 higher  - quiere cursar estudios superiores (binario: sí o no)
- 22 internet - Acceso a internet en casa (binario: si o no)
- 23 romantic  - con una relación romántica (binario: sí o no)
- 24 famrel - calidad de las relaciones familiares (numérico: de 1 - muy mala a 5 excelente)
- 25 freetime  - tiempo libre después de la escuela (numérico: de 1 - muy bajo a 5 - muy alto)
- 26 goout - salir con amigos (numérico: de 1 - muy bajo a 5 - muy alto)
- 27 Dalc - consumo de alcohol en jornada laboral (numérico: de 1 - muy bajo a 5 - muy alto)
- 28 Walc - consumo de alcohol en fin de semana (numérico: de 1 - muy bajo a 5 - muy alto)
- 29 health  - estado de salud actual (numérico: de 1 - muy malo a 5 - muy bueno)
- 30 absences  - número de ausencias escolares (numérico: de 0 a 93)

## **Estas calificaciones están relacionadas con la materia del curso, Matemáticas o Portugués:**

- 31 G1 - nota del primer período (numérico: de 0 a 20)
- 31 G2 - nota del segundo período (numérico: de 0 a 20)
- 32 G3 - nota final (numérico: de 0 a 20, objetivo de salida)

## **a. Recordemos nuevamente nuestras columnas**

In [ ]:
## **a. Recordemos nuevamente nuestras columnas**
data3.columns

Index(['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2',
       'G3'],
      dtype='object')

## **b. Usaremos el modulo de Sklearn para imputar los datos faltantes con la media**
- Dado que las variables con valores perdidos se encuentra todas en la escala Liker, es conveniente imputarlas con el valor que corresponda al 50% de los datos y generalmente es la mediana.

In [ ]:
## **b. Usaremos el modulo de Sklearn para imputar los datos faltantes con la media**
from sklearn.impute import SimpleImputer # llamamos la función que vamos a necesitar
imp = SimpleImputer(missing_values=np.nan, strategy='median') # vamos a cambiar los valores perdidos usando la estaretgia media
imp.fit(data3) # la función fit del objeto creado y sacar la media
data3 = imp.transform(data3) # me esta transformando los datos como queremos
# data3 si imprimimos aqui, en este paso hace los cambios pero lo devuelve como un arreglo
data3 = pd.DataFrame(imp.transform(data3), columns =['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel',
       'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2',
       'G3'])
data3


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but SimpleImputer was fitted with feature names
  warnings.warn(


age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  \
0    18.0   3.0   4.0         1.0        2.0       0.0     4.0       3.0   
1    17.0   3.0   3.0         1.0        2.0       0.0     5.0       3.0   
2    17.0   1.0   3.0         1.0        2.0       3.0     4.0       3.0   
3    15.0   4.0   2.0         1.0        2.0       0.0     3.0       3.0   
4    17.0   3.0   3.0         1.0        2.0       0.0     4.0       3.0   
..    ...   ...   ...         ...        ...       ...     ...       ...   
390  20.0   2.0   3.0         1.0        2.0       2.0     4.0       3.0   
391  17.0   3.0   1.0         2.0        1.0       0.0     2.0       4.0   
392  21.0   3.0   1.0         1.0        2.0       0.0     5.0       5.0   
393  18.0   3.0   3.0         1.0        2.0       0.0     4.0       4.0   
394  19.0   1.0   1.0         1.0        1.0       0.0     3.0       2.0   

     goout  Dalc  Walc  health  absences    G1    G2    G3  
0      4.0   1.0   1.0     3.0       6.0   5.0   6.0   6.0  
1      3.0   1.0   1.0     3.0       4.0   5.0   5.0   6.0  
2      3.0   1.0   3.0     3.0      10.0   7.0   8.0  10.0  
3      3.0   1.0   1.0     5.0       2.0  15.0  14.0  15.0  
4      2.0   1.0   2.0     5.0       4.0   6.0  10.0  10.0  
..     ...   ...   ...     ...       ...   ...   ...   ...  
390    4.0   4.0   2.0     4.0      11.0   9.0   9.0   9.0  
391    3.0   3.0   4.0     2.0       3.0  14.0  16.0  16.0  
392    3.0   3.0   2.0     3.0       3.0  10.0   8.0   7.0  
393    1.0   3.0   2.0     5.0       0.0  11.0  12.0  10.0  
394    3.0   3.0   3.0     5.0       5.0   8.0   9.0   9.0  

[395 rows x 16 columns]

In [ ]:
## **c. Podemos verificar que tan buena fue nuestra imputación**
verificacion = data0-data3

In [ ]:
verificacion.describe()

age        Medu        Fedu  traveltime  studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.00000  395.000000   
mean    -0.126582   -0.037975   -0.215190    0.182278    0.04557    0.200000   
std      0.663919    0.649932    0.751832    0.480260    0.61119    0.590105   
min     -2.000000   -3.000000   -3.000000    0.000000   -1.00000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.00000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.00000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.00000    0.000000   
max      3.000000    1.000000    1.000000    3.000000    2.00000    3.000000   

         famrel    freetime       goout        Dalc        Walc      health  \
count  395.0000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     0.0000    0.065823    0.065823    0.116456    0.116456   -0.017722   
std      0.2363    0.600948    0.815909    0.479644    0.791200    0.337488   
min     -3.0000   -2.000000   -2.000000    0.000000   -1.000000   -3.000000   
25%      0.0000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.0000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.0000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.0000    2.000000    2.000000    4.000000    3.000000    1.000000   

         absences     G1     G2     G3  
count  395.000000  395.0  395.0  395.0  
mean     0.136709    0.0    0.0    0.0  
std      2.043075    0.0    0.0    0.0  
min     -4.000000    0.0    0.0    0.0  
25%      0.000000    0.0    0.0    0.0  
50%      0.000000    0.0    0.0    0.0  
75%      0.000000    0.0    0.0    0.0  
max     16.000000    0.0    0.0    0.0